crates/interpreter/src/interpreter/shared_stack.rs

```diff
@@ -0,0 +1,775 @@
+    /// Invariant: it a valid pointer to `self.pages[self.page_idx].buffer`
+    buffer: *mut Buffer,
+    /// Keeps track of the length of the stack in the current context.
+    /// Needed for better performance and to avoid double
+    /// heap lookup for basic stack operations.
+    ///
+    /// Invariant: it a valid pointer to `self.pages[self.page_idx].buffer`
+    context_len: usize,
+}
+
+impl fmt::Display for SharedStack {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("[")?;
+        for (i, x) in self.context_stack().iter().enumerate() {
+            if i > 0 {
+                f.write_str(", ")?;
+            }
+            write!(f, "{x}")?;
+        }
+        f.write_str("]")
+    }
+}
+
+impl Default for SharedStack {
+    #[inline]
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl SharedStack {
+    /// Instantiate a new shared stack with a single page of size `PAGE_SIZE`
+    #[inline]
+    pub fn new() -> Self {
+        let mut pages = vec![Page::new()];
+        let buffer = unsafe { &mut pages.get_unchecked_mut(0).buffer };
+        Self {
+            buffer,
+            pages,
+            page_idx: 0,
+            context_len: 0,
+        }
+    }
+
+    /// Prepares the shared stack for a new context
+    #[inline]
+    pub fn new_context(&mut self) {
+        // check memory left for the current page
+        let page = self.page_mut();
+        let memory_left = page.buffer.capacity() - page.buffer.len();
+
+        if memory_left < STACK_LIMIT {
+            // we need to move to the next page
+            self.page_idx += 1;
+            if self.pages.len() == self.page_idx {
+                // new page needed
+                self.pages.push(Page::new());
+            }
+            let page = self.page_mut();
+            page.checkpoints.push(0);
+            self.buffer = &mut page.buffer
+        } else {
+            page.checkpoints.push(page.buffer.len())
+        }
+        self.context_len = 0;
+    }
+
+    /// Prepares the shared stack for returning to the previous context
+    #[inline]
+    pub fn free_context(&mut self) {
+        let page = self.page_mut();
+        // SAFETY: `new_context` logic always guarantees at least the zero checkpoint in it
+        let old_checkpoint = unsafe { page.checkpoints.pop().unwrap_unchecked() };
+        if let Some(last_checkpoint) = page.checkpoints.last() {
+            // SAFETY: checkpoints are always in bound of buffers
+            unsafe {
+                page.buffer.set_len(old_checkpoint);
+            };
+            self.context_len = old_checkpoint - last_checkpoint;
+        } else if self.page_idx > 0 {
+            // no more checkpoints means we need to move to the previous page
+            self.page_idx -= 1;
+            self.buffer = &mut self.page_mut().buffer;
+            // SAFETY: Going back to previous page implies non-empty checkpoints list
+            self.context_len = unsafe {
```

Comment:
Why is this unsafe?

Reply:
Because of the `unwrap_unchecked` at the end of line 144. When going back to the previous page checkpoints length is greater than zero by construction, therefore checking the `Option<T>` is not needed.

Am I missing something?

Reply:
brainfart, ignore it I just saw `unwrap` and not `unwrap_unchecked`